src/main/java/com/nirima/jenkins/plugins/docker/DockerContainerWatchdog.java

```diff
@@ -0,0 +1,538 @@
+import java.time.Instant;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.dockerjava.api.DockerClient;
+import com.github.dockerjava.api.model.Container;
+import com.nirima.jenkins.plugins.docker.utils.JenkinsUtils;
+
+import hudson.Extension;
+import hudson.model.AsyncPeriodicWork;
+import hudson.model.Node;
+import hudson.model.TaskListener;
+import hudson.slaves.Cloud;
+import hudson.slaves.SlaveComputer;
+import io.jenkins.docker.DockerTransientNode;
+import io.jenkins.docker.client.DockerAPI;
+import jenkins.model.Jenkins;
+
+/**
+ * Periodic job, which gets executed by Jenkins automatically, to ensure the
+ * consistency of the containers currently running on the docker and the nodes
+ * which are attached to this Jenkins instance.
+ * 
+ * @author eaglerainbow
+ *
+ */
+
+@Extension
+public class DockerContainerWatchdog extends AsyncPeriodicWork {
+    private Clock clock;
+    
+    public DockerContainerWatchdog() {
+        super(String.format("%s Asynchronous Periodic Work", DockerContainerWatchdog.class.getSimpleName()));
+        this.clock = Clock.systemUTC();
+    }
+
+    /**
+     * sets the internal clock of this class - only to be used in unit testing environment!
+     * @param clock the clock which shall be used from now on
+     */
+    @Restricted(NoExternalUse.class)
+    public void setClock(Clock clock) {
+        this.clock = clock;
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DockerContainerWatchdog.class);
+
+    /**
+     * The recurrence period how often this task shall be run
+     */
+    private static final long RECURRENCE_PERIOD_IN_MS = JenkinsUtils.getSystemPropertyLong(DockerContainerWatchdog.class.getName()+".recurrenceInSeconds", Long.valueOf(5*60))*1000L;
+
+    /**
+     * The duration, which defines the maximal amount of time the watchdog is allowed to run 
+     * for a single execution.
+     * The processing will be stopped on next occasion, if this threshold is exceeded.
+     */
+    private static final Duration PROCESSING_TIMEOUT_IN_MS = Duration.ofMillis(RECURRENCE_PERIOD_IN_MS * 4 / 5);
+
+    /**
+     * The duration, which is permitted containers to start/run without having a node attached.
+     * This is to prevent that the watchdog may be undesirably kill containers, which are just
+     * on the way to the be started.
+     * It automatically also is a "minimal lifetime" value for containers, before this watchdog
+     * is allowed to kill any container. 
+     */
+    private static final Duration GRACE_DURATION_FOR_CONTAINERS_TO_START_WITHOUT_NODE_ATTACHED = Duration.ofSeconds(JenkinsUtils.getSystemPropertyLong(DockerContainerWatchdog.class.getName()+".initialGraceDurationForContainersInSeconds", 60L));
+    
```

Comment:
General code formatting issue:  Could you remove trailing whitespace from all lines of code being added or changed in this PR?
e.g. Do a regex search-and-replace to replace "space space asterisk dollar" with the empty string?

It's not _purely_ [OCD](https://en.wikipedia.org/wiki/Obsessive%E2%80%93compulsive_disorder) ;-) - it can cause merge issues when people accidentally add or remove whitespace (e.g. their IDE does this without their knowledge) because it makes git (etc) think that lines have had important changes made when they haven't.

Reply:
Done with 3920fd0e8c43f4b4f4c19f96ac2c14a06a9a9af0 globally (on all files changed in this PR). 
Normally, my auto-formatter in the IDE does that automatically, but I don't use it here anymore, as you then would start to scream at other cases it does :stuck_out_tongue: